src/keychain/index.js

```diff
@@ -503,6 +506,60 @@ class Keychain {
       return throwDelayed(errcode(new Error(`Key '${name}' does not exist. ${err.message}`), 'ERR_KEY_NOT_FOUND'))
     }
   }
+
+  /**
+   * Rotate keychain password and re-encrypt all assosciated keys
+   *
+   * @param {string} oldPass - The old local keychain password
+   * @param {string} newPass - The new local keychain password
+   */
+  async rotateKeychainPass (oldPass, newPass) {
+    if (typeof oldPass !== 'string') {
+      return throwDelayed(errcode(new Error(`Invalid old pass type '${typeof oldPass}'`), 'ERR_INVALID_OLD_PASS_TYPE'))
+    }
+    if (typeof newPass !== 'string') {
+      return throwDelayed(errcode(new Error(`Invalid new pass type '${typeof newPass}'`), 'ERR_INVALID_NEW_PASS_TYPE'))
+    }
+    if (newPass.length < 20) {
+      return throwDelayed(errcode(new Error(`Invalid pass length ${newPass.length}`), 'ERR_INVALID_PASS_LENGTH'))
+    }
+    log('recreating keychain')
+    const oldDek = privates.get(this).dek
+    this.opts.pass = newPass
+    const newDek = newPass
+      ? crypto.pbkdf2(
+        newPass,
+        this.opts.dek.salt,
+        this.opts.dek.iterationCount,
+        this.opts.dek.keyLength,
+        this.opts.dek.hash)
+      : ''
+    privates.set(this, { dek: newDek })
+    const keys = await this.listKeys()
+    for (const key of keys) {
+      const res = await this.store.get(DsName(key.name))
+      const pem = uint8ArrayToString(res)
+      const privateKey = await crypto.keys.import(pem, oldDek)
+      const password = newDek.toString()
+      const keyAsPEM = await privateKey.export(password)
+
+      // Remove key with old pass
+      const batch = this.store.batch()
+      batch.delete(DsName(key.name))
+      batch.delete(DsInfoName(key.name))
+      await batch.commit()
+
```

Comment:
```suggestion
```

The last commit should be enough right?

Reply:
I thought so too at first but I believe there's some sort of issue where batch cannot both delete and create a key of the same name without separate commits. When I remove the first `batch.commit()` and run `rotateKeychainPass()`, the datastore appears to be completely empty like so: `MemoryDatastore { data: {} }`. If I try `batch.put()` with a different key name, the operation seems to work as expected (i.e the old key gets deleted and the new key is appended). 

Try running the below and let me know If I'm missing something or if this is a legitimate issue:
```js
await kc.createKey('keyCreatedWithOldPassword', 'rsa', 2048)
const dsnameData = new Key('/pkcs8/keyCreatedWithOldPassword')
const dsnameInfo = new Key('/info/keyCreatedWithOldPassword')
const batch = ds.batch()

// Delete key data and info
batch.delete(dsnameData)
batch.delete(dsnameInfo)

// Append key data and info
await batch.put(dsnameData, uint8ArrayFromString('pem'))
await batch.put(dsnameInfo, uint8ArrayFromString('info'))

await batch.commit()
console.log(ds) // Output: MemoryDatastore { data: {} }
```

Reply:
Thanks for the details. I see, but rethinking this. A datastore will always be indexed by a key, so the `delete` is redundant as the `put` will write over it

Reply:
Added two suggestions to get the delete out again.